lib/luacrypto_aead.c

```diff
@@ -0,0 +1,391 @@
+/*
+* SPDX-FileCopyrightText: (c) 2025 jperon <cataclop@hotmail.com>
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+/***
+Low-level Lua interface to the Linux Kernel Crypto API for AEAD
+(Authenticated Encryption with Associated Data) ciphers.
+ *
+This module provides a `new` function to create AEAD transform objects,
+which can then be used for encryption and decryption.
+@see crypto_aead_tfm
+
+@module crypto_aead
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/aead.h>
+#include <linux/err.h>
+#include <linux/scatterlist.h>
+#include <linux/slab.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+static const lunatik_class_t luacrypto_aead_tfm_class;
+
+typedef struct {
+	struct crypto_aead *tfm;
+	struct aead_request *req;
+
+	u8 *iv_data;
+	size_t iv_len;
+
+	u8 *work_buffer;
+	size_t work_buffer_len;
+
+	size_t aad_len_in_buffer;
+	size_t crypt_data_len_in_buffer; // For encrypt: PT len. For decrypt: CT+Tag len.
+
+	struct scatterlist sg_work;
+} luacrypto_aead_tfm_t;
+
+LUNATIK_PRIVATECHECKER(luacrypto_check_aead_tfm, luacrypto_aead_tfm_t *);
+
+static void luacrypto_aead_tfm_release(void *private)
+{
+	luacrypto_aead_tfm_t *tfm_ud = (luacrypto_aead_tfm_t *)private;
+	if (!tfm_ud) {
+		return;
+	}
+
+	kfree(tfm_ud->iv_data);
+	tfm_ud->iv_data = NULL;
+	kfree(tfm_ud->work_buffer);
+	tfm_ud->work_buffer = NULL;
+
+	if (tfm_ud->req) {
+		aead_request_free(tfm_ud->req);
+		tfm_ud->req = NULL;
+	}
+
+	if (tfm_ud->tfm && !IS_ERR(tfm_ud->tfm)) {
+		crypto_free_aead(tfm_ud->tfm);
+	}
+}
+
+
+/// AEAD Transform (TFM) methods.
```

Comment:
we are not using cpp comment style.. we try to keep something close to the kernel code base just to reduce the friction.. =)